cloud/ec2_price.go

```diff
@@ -93,85 +94,186 @@ func getZone(ctx context.Context, client AWSClient, h *host.Host) (string, error
+	if err != nil {
+		return 0, errors.Wrap(err, "problem querying for pricing data")
+	}
 
-	details := struct {
-		Terms    Terms
-		Products map[string]struct {
-			SKU           string
-			ProductFamily string
-			Attributes    struct {
-				Location        string
-				InstanceType    string
-				PreInstalledSW  string
-				OperatingSystem string
-				Tenancy         string
-				LicenseModel    string
-			}
-		}
-	}{}
+	p, err := cpf.parseAWSPricing(out)
+	if err != nil {
+		return 0, errors.Wrap(err, "problem parsing aws pricing")
+	}
+	cpf.Lock()
+	defer cpf.Unlock()
+	cpf.ec2Prices[odInfo{os: osPriceName, instance: instance, region: region}] = p
+	return p, nil
+}
 
-	_, err := util.Retry(func() (bool, error) {
-		resp, err := client.Get(endpoint)
-		if resp != nil {
-			defer resp.Body.Close()
+func (cpf *cachingPriceFetcher) makeGetProductsInput(info odInfo) *pricing.GetProductsInput {
+	const match = "TERM_MATCH"
+	constructGetProductsInput := &pricing.GetProductsInput{
+		Filters: []*pricing.Filter{
+			{
+				Field: aws.String("ServiceCode"),
+				Type:  aws.String(match),
+				Value: aws.String("AmazonEC2"),
+			},
+			{
+				Field: aws.String("productFamily"),
+				Type:  aws.String(match),
+				Value: aws.String("Compute Instance"),
+			},
+			{
+				Field: aws.String("preInstalledSw"),
+				Type:  aws.String(match),
+				Value: aws.String("NA"),
+			},
+			{
+				Field: aws.String("tenancy"),
+				Type:  aws.String(match),
+				Value: aws.String("Shared"),
+			},
+			{
+				Field: aws.String("capacityStatus"),
+				Type:  aws.String(match),
+				Value: aws.String("UnusedCapacityReservation"),
+			},
+			{
+				Field: aws.String("instanceType"),
+				Type:  aws.String(match),
+				Value: aws.String(info.instance),
+			},
+			{
+				Field: aws.String("operatingSystem"),
+				Type:  aws.String(match),
+				Value: aws.String(info.os),
+			},
+			{
+				Field: aws.String("location"),
+				Type:  aws.String(match),
+				Value: aws.String(info.region),
+			},
+		},
+		ServiceCode: aws.String("AmazonEC2"),
+	}
+	return constructGetProductsInput
+}
+
+// Parse a JSON document like this one
+// {
+// 	"version": "20181031070014",
+// 	"terms": {
+// 		"OnDemand": {
+// 			"XCH2WN4F4MYH63N8.JRTCKXETXF": {
+// 				"termAttributes": {
+// 				},
+// 				"sku": "XCH2WN4F4MYH63N8",
+// 				"priceDimensions": {
+// 					"XCH2WN4F4MYH63N8.JRTCKXETXF.6YS6EN2CT7": {
+// 						"unit": "Hrs",
+// 						"rateCode": "XCH2WN4F4MYH63N8.JRTCKXETXF.6YS6EN2CT7",
+// 						"pricePerUnit": {
+// 							"USD": "0.8400000000"
+// 						},
+// 						"endRange": "Inf",
+// 						"description": "$0.840 per Unused Reservation Linux c3.4xlarge Instance Hour",
+// 						"beginRange": "0",
+// 						"appliesTo": []
+// 					}
+// 				},
+// 				"offerTermCode": "JRTCKXETXF",
+// 				"effectiveDate": "2018-10-01T00:00:00Z"
+// 			}
+// 		}
+// 	},
+// [...]
+func (cpf *cachingPriceFetcher) parseAWSPricing(out *pricing.GetProductsOutput) (float64, error) {
```

Comment:
Given that you have to do something ugly like this, does it make more sense to marshal it to a string and string match?

Reply:
I considered that. It's true that this is ugly, but since the data is structured, and since the string "USD" may appear elsewhere (there's a plaintext field with a phrase describing pricing) I opted for the uglier structured data parsing. Our integration test hits the actual API ,and I've hardcoded actual on-demand prices into it, so it should fail if anything changes about their data structure.